src/chat.cpp

```diff
@@ -912,13 +980,31 @@ void CHAT_PrintChatString( ULONG ulPlayer, ULONG ulMode, const char *pszString )
 			S_Sound( CHAN_VOICE | CHAN_UI, "misc/chat", 1, ATTN_NONE );
 	}
 
-	BOTCMD_SetLastChatString( pszString );
-	BOTCMD_SetLastChatPlayer( PLAYER_IsValidPlayer( ulPlayer ) ? players[ulPlayer].userinfo.GetName() : "" );
-
+	// [AK] Only save chat messages for non-private chat messages.
+	if (( ulMode != CHATMODE_PRIVATE_SEND ) && ( ulMode != CHATMODE_PRIVATE_RECEIVE ))
 	{
-		ULONG	ulIdx;
+		if ( ulPlayer == MAXPLAYERS )
+		{
+			SERVER_AddChatMessage( pszString );
+		}
+		else
+		{
+			FString message = pszString;
 
-		for ( ulIdx = 0; ulIdx < MAXPLAYERS; ulIdx++ )
+			// [AK] If there's too many stored messages, remove the oldest one.
+			if ( players[ulPlayer].ChatMessages.Size() >= MAX_STORED_MESSAGES )
+				players[ulPlayer].ChatMessages.Delete( MAX_STORED_MESSAGES - 1 );
+
+			players[ulPlayer].ChatMessages.Insert( 0, message );
```

Comment:
A `TArray` is not a very efficient data structure for this, because you are copying the entire existing content of the TArray when you add something in front. For this kind of use, a ring buffer would be much more efficient. I think you can use `RingBuffer` from `networkshared.h` instead here. It's designed for keeping the last `N` entries. You may just have to generalize `RingBuffer::clear()`, since it was not created for `DataType`'s that are classes with their own memory management (like `FString`).

Reply:
I see, thank you for the suggestion. Would changing the data type from `FString` to `char *` work better with the ring buffer?

Reply:
Yeah, sticking with an FString should be the easiest. I think the only thing necessary should be to call the destructor on each stored `FString` in `RingBuffer::clear()`, i.e. only the line
```memset( _data, 0, sizeof( _data ))```
needs adjustment.